prql-java/prql-api/src/main/rust/src/lib.rs

```diff
@@ -0,0 +1,75 @@
+use std::str::FromStr;
+
+use jni::objects::{JClass, JObject, JString};
+use jni::sys::jstring;
+use jni::JNIEnv;
+use prql_compiler::sql::Dialect;
+use prql_compiler::{json, prql_to_pl, Options, Target};
+
+// exception handling: https://github.com/jni-rs/jni-rs/issues/154
+
+#[no_mangle]
+#[allow(non_snake_case)]
+pub extern "system" fn Java_org_prqllang_internal_JniImpl_compile0<'local>(
+    mut env: JNIEnv<'local>,
+    _class: JClass<'local>,
+    query: JString<'local>,
+    option: JObject<'local>,
+) -> jstring {
+    let prql_query: String = env
+        .get_string(&query)
+        .expect("Couldn't get java string!")
+        .into();
+
+    let java_target: JString = env
+        .call_method(&option, "getTarget", "()Ljava/lang/String;", &[])
+        .unwrap()
+        .l()
+        .unwrap()
+        .into();
+    let rust_target: String = env
+        .get_string(&java_target)
+        .expect("Couldn't get java string!")
+        .into();
+    let target = Target::Sql(Some(Dialect::from_str(&rust_target).unwrap_or_default()));
+    let format = env
+        .call_method(&option, "requireFormat", "()Z", &[])
+        .unwrap()
+        .z()
+        .unwrap();
+    let signature_comment = env
+        .call_method(&option, "requireComment", "()Z", &[])
+        .unwrap()
+        .z()
+        .unwrap();
+    let option = Options {
+        format,
+        target,
+        signature_comment,
+    };
+
+    let rs_sql_str: String =
+        prql_compiler::compile(&prql_query, &option).expect("Couldn't compile query to SQL!");
```

Comment:
Re your question on handling errors — my guess is that here we need to instruct it to throw the error in Java with something like (from your link): https://github.com/jni-rs/jni-rs/issues/154#issuecomment-446681627

Reply:
Yes, I'm asking for your help. I brought a book and just started to read tutorials created for Java developers, but it's going to take a while for me enhance the code. Appreciate if you can take care of changes on rust side, thanks in advance!

Reply:
OK, I can likely have a look soon. But I'll need to get a Java environment set up, and I do most of my coding on the weekend. 

In the meantime, I think something like this should at least raise a simple error:

```suggestion
        prql_compiler::compile(&prql_query, &option).unwrap_or(env.throw_new("java/lang/RuntimeException", "Query error (TODO a better description)"));
```

Reply:
Thanks but it generates compile error.

Reply:
@zhicwu this is one approach — I'm not sure it's great, but it works OK:

```diff
diff --git a/prql-java/prql-api/src/main/rust/src/lib.rs b/prql-java/prql-api/src/main/rust/src/lib.rs
index c711961..58c67d9 100644
--- a/prql-java/prql-api/src/main/rust/src/lib.rs
+++ b/prql-java/prql-api/src/main/rust/src/lib.rs
@@ -48,11 +48,23 @@ pub extern "system" fn Java_org_prqllang_internal_JniImpl_compile0<'local>(
         signature_comment,
     };
 
-    let rs_sql_str: String =
-        prql_compiler::compile(&prql_query, &option).expect("Couldn't compile query to SQL!");
-    env.new_string(rs_sql_str)
-        .expect("Couldn't create java string!")
-        .into_raw()
+    match prql_compiler::compile(&prql_query, &option) {
+        Ok(sql) => env
+            .new_string(sql)
+            .expect("Couldn't create java string!")
+            .into_raw(),
+        Err(e) => {
+            env.throw_new(
+                "java/lang/RuntimeException",
+                format!("Query error: {}", e).as_str(),
+            )
+            .unwrap();
+
+            // TODO: can we return something more meaningful than an empty
+            // string? Are there better docs than https://github.com/jni-rs/jni-rs/issues/154?
+            env.new_string("").unwrap().into_raw()
+        }
+    }
 }
 
 #[no_mangle]
```